packages/server/postgres/migrations/2025-04-17T13:33:36.200Z_RBAC.ts

```diff
@@ -0,0 +1,261 @@
+    UNION ALL
+    SELECT poa.role
+    FROM "PageOrganizationAccess" poa
+    JOIN "OrganizationUser" ou ON poa."orgId" = ou."orgId"
+    WHERE ou."userId" = "_userId" AND poa."pageId" = "_pageId"
+  ) AS effective_roles;
+
+  -- Check existing effective role
+  SELECT role INTO "currentRole"
+  FROM "PageAccess"
+  WHERE "userId" = "_userId" AND "pageId" = "_pageId";
+
+  IF "strongestRole" IS NULL THEN
+    -- User lost access
+    DELETE FROM "PageAccess"
+    WHERE "userId" = "_userId" AND "pageId" = "_pageId";
+  ELSIF "currentRole" IS NULL THEN
+    -- New access
+    INSERT INTO "PageAccess" ("userId", "pageId", role)
+    VALUES ("_userId", "_pageId", "strongestRole");
+  ELSIF "strongestRole" <> "currentRole" THEN
+    -- Changed access
+    UPDATE "PageAccess"
+    SET role = "strongestRole"
+    WHERE "userId" = "_userId" AND "pageId" = "_pageId";
+  END IF;
+END;
+$$ LANGUAGE plpgsql;
+
+
+-- UPDATE CACHE FROM PageUserAccess
+CREATE OR REPLACE FUNCTION "updateUserPageAccess"() RETURNS TRIGGER AS $$
+BEGIN
+  PERFORM "updatePageAccess"(NEW."userId", NEW."pageId");
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE TRIGGER trg_user_page_access
+AFTER INSERT OR UPDATE OR DELETE ON "PageUserAccess"
+FOR EACH ROW EXECUTE FUNCTION "updateUserPageAccess"();
+
+
+-- UPDATE CACHE FROM PageTeamAccess
+CREATE OR REPLACE FUNCTION "updateTeamPageAccess"() RETURNS TRIGGER AS $$
+DECLARE
+  "_userId" VARCHAR;
+BEGIN
+  FOR "_userId" IN SELECT "userId" FROM "TeamMember" WHERE "teamId" = NEW."teamId" LOOP
+  PERFORM "updatePageAccess"("_userId", NEW."pageId");
+  END LOOP;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE TRIGGER trg_team_page_access
+AFTER INSERT OR UPDATE OR DELETE ON "PageTeamAccess"
+FOR EACH ROW EXECUTE FUNCTION "updateTeamPageAccess"();
+
+
+-- UPDATE CACHE FROM PageOrganizationAccess
+CREATE OR REPLACE FUNCTION "updateOrganizationPageAccess"() RETURNS TRIGGER AS $$
+DECLARE
+  "_userId" VARCHAR;
+BEGIN
+  FOR "_userId" IN SELECT "userId" FROM "OrganizationUser" WHERE "orgId" = NEW."orgId" LOOP
+  PERFORM "updatePageAccess"("_userId", NEW."pageId");
+  END LOOP;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE TRIGGER trg_org_page_access
+AFTER INSERT OR UPDATE OR DELETE ON "PageOrganizationAccess"
+FOR EACH ROW EXECUTE FUNCTION "updateOrganizationPageAccess"();
+
+
+-- HANDLE JOIN/LEAVE TEAM
+CREATE OR REPLACE FUNCTION "updateTeamPageAccessByTeamMember"() RETURNS TRIGGER AS $$
+DECLARE
+  "_pageId" INT;
+BEGIN
+  IF (TG_OP = 'INSERT') OR (TG_OP = 'UPDATE' AND NEW."isNotRemoved" != OLD."isNotRemoved") THEN
+    FOR "_pageId" IN SELECT "pageId" FROM "PageTeamAccess" WHERE "teamId" = NEW."teamId" LOOP
+      PERFORM "updatePageAccess"(NEW."userId", "_pageId");
+    END LOOP;
+  END IF;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE TRIGGER trg_team_member_update_team_page_access
+AFTER INSERT OR UPDATE ON "TeamMember"
```

Comment:
+1 Are we sure we never delete team members, even manually? I would feel better to have a trigger for it.